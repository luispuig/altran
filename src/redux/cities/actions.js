// @flow

import openWeather from "../../services/openWeather";

import type { _RequestPayload, _SuccesPayload, _FailPayload } from "./types.js"
import type { _State } from "../types.js"


export const CITY_UPDATE__REQUEST: string = "CITY_UPDATE__REQUEST";
export const CITY_UPDATE__FAIL: string = "CITY_UPDATE__FAIL";
export const CITY_UPDATE__SUCCESS: string = "CITY_UPDATE__SUCCESS";

export const cityUpdateAll = () => {

  /*
    ApiWeather allows to get all cities in one request with a limit of 20.
    But each city will be fetched individually for showing a more standard way.
    This way will allow future features like different update time per city.
   */

  return (dispatch:any => void, getState:any => void ) => {

    const state:_State = getState();
    const { cities } = state;
    // Dispatch Request Action for each City
    cities.map( item => dispatch( cityUpdate( { id: item.id } ) ) );
  }

};

export const cityUpdate = ( { id }:_RequestPayload, date:?Date = undefined ) => {
  return (dispatch:any => void) => {

    // Set City State to Fetching Request
    dispatch( cityUpdate_Request( { id } ) );

    // Fetch Promise
    return openWeather.getTempByCityId( id ).then(
        (temperature:number) => {
          dispatch( cityUpdate_Success  ( { id, temperature } ) );
        },
        (error:{ message:string }) => {
          dispatch( cityUpdate_Fail( { id, error: error.message } ) ) }
    );
  }
};

export const cityUpdate_Request = ({ id }:_RequestPayload) => ({
  type: CITY_UPDATE__REQUEST,
  payload: { id },
});

export const cityUpdate_Fail = ({id, error}:_FailPayload) => ({
  type: CITY_UPDATE__FAIL,
  payload: {id, error}
});

export const cityUpdate_Success = ({id, temperature}:_SuccesPayload) => ({
  type: CITY_UPDATE__SUCCESS,
  payload: {id, temperature}
});
