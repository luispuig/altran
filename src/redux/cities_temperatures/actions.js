// @flow
import type {_AddPayload} from "./types"

export const CITY_TEMPERATURE__ADD:string = "CITY_TEMPERATURE__ADD";

export const cityTemperature_Add = ({id, temperature, date}:{..._AddPayload, date:?Date}) => ({
  type: CITY_TEMPERATURE__ADD,
  payload: { id, temperature, date: (date !== undefined)? date: new Date() }
});
