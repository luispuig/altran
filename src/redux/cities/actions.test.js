import * as actions from "./actions";

it('Creates Actions cityUpdateAll', async () => {
  const response = { main: { temp: 25 } };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  const action = actions.cityUpdateAll();
  const dispatchMock = jest.fn() ;
  const initialData = { temperature: 0, state: 'request', error: undefined  };
  const cities = [
    { id: 3871336, name: "Santiago",      ...initialData },
    { id: 3435910, name: "Buenos Aires",  ...initialData },
    { id: 3936456, name: "Lima",          ...initialData },
    { id: 3448439, name: "Sao Paulo",     ...initialData }
  ];

  // Dispach Actions and Get Array of Action Creators
  action(dispatchMock, () => ({cities}) );

  const dispatchMock2 = jest.fn() ;
  dispatchMock.mock.calls.map( async actionCreator => {
    const fn = actionCreator[0];
    //console.log(fn.toString());
    // Call each cityUpdate action creator
    fn(dispatchMock2);

  });

  // Result of calling every cityUpdate action creator
  const finalActions = dispatchMock2.mock.calls;

  expect(finalActions).toMatchSnapshot();
});


it('Creates Action cityUpdate', async () => {

  const response = { main: { temp: 25 } };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  const action = actions.cityUpdate({ id: 3871336}, new Date('2017-12-06T14:14:58.080Z'));
  const dispatchMock = jest.fn() ;

  // Dispach Actions
  const apiCallPromise = action(dispatchMock);
  await apiCallPromise;

  expect(dispatchMock.mock.calls).toMatchSnapshot();

});

it('Creates Action cityUpdate-ERROR', async () => {

  const response = { cod: 404, message: "error" };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  const action = actions.cityUpdate({ id: 3871336}, new Date('2017-12-06T14:14:58.080Z'));
  const dispatchMock = jest.fn() ;

  // Dispach Actions
  const apiCallPromise = action(dispatchMock);
  await apiCallPromise;

  expect(dispatchMock.mock.calls).toMatchSnapshot();

});


it('Creates Action cityUpdate_Request', () => {
  const action = actions.cityUpdate_Request({ id: 123 })
  expect(action).toMatchSnapshot();
});

it('Creates Action cityUpdate_Fail', () => {
  const action = actions.cityUpdate_Fail({ id: 123, error: "error message" })
  expect(action).toMatchSnapshot();
});

it('Creates Action cityUpdate_Success', () => {
  const action = actions.cityUpdate_Success({ id: 123, temperature:15 })
  expect(action).toMatchSnapshot();
});
