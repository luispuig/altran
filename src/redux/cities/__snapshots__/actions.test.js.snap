// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates Action cityUpdate 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "id": 3871336,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": 3871336,
        "temperature": 25,
      },
      "type": "CITY_UPDATE__SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "date": 2017-12-06T14:14:58.080Z,
        "id": 3871336,
        "temperature": 25,
      },
      "type": "CITY_TEMPERATURE__ADD",
    },
  ],
]
`;

exports[`Creates Action cityUpdate_Fail 1`] = `
Object {
  "payload": Object {
    "error": "error message",
    "id": 123,
  },
  "type": "CITY_UPDATE__FAIL",
}
`;

exports[`Creates Action cityUpdate_Request 1`] = `
Object {
  "payload": Object {
    "id": 123,
  },
  "type": "CITY_UPDATE__REQUEST",
}
`;

exports[`Creates Action cityUpdate_Success 1`] = `
Object {
  "payload": Object {
    "id": 123,
    "temperature": 15,
  },
  "type": "CITY_UPDATE__SUCCESS",
}
`;

exports[`Creates Action cityUpdate-ERROR 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "id": 3871336,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "error": "error",
        "id": 3871336,
      },
      "type": "CITY_UPDATE__FAIL",
    },
  ],
]
`;

exports[`Creates Actions cityUpdateAll 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "id": 3871336,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": 3435910,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": 3936456,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": 3448439,
      },
      "type": "CITY_UPDATE__REQUEST",
    },
  ],
]
`;
